### ISSUE TITLE
How to use pytorch's optimization method

### ISSUE THREAD

**Issue by qiansichong** [lixilinx/IVA4Cocktail#3]
Hi,

In your paper or code, the density model parameters Ws is updated using Preconditioned SGD optimizer( this SGD algorithm is very interesting). If I want to use pytorch's  optimization method, such as torch.optim.Adam( ), what should I do?
What I don’t understand is that there is still an updating of W_iva between density model and  loss. How to do its backpropagation?

**Comment by lixilinx** [comment]
It should be easy to apply Pytorch's on shelf optimizer like adam. You can also do it in a few lines of code. Just note that 
grads = grad(loss, Ws)
is the gradient. It is all the information required by adam. 

The gradients are of very small amplitudes. So the typical parameter settings of adam perform poorly in my trials. You may try to find a good set of  such optimization related parameters settings for adam. I used psgd, and one main reason is that I have no need to tune those optimization related hyper parameters since everything is normalized. Definitely, psgd is not necessarily the best choice. 

The last question is somewhat nontrivial. As you can see, the density model determines everything else like the update of W_iva, separation results, etc. Basically, W_iva[0] --> W_iva[1] --> W_iva[2] .... This defines an extremely deep recurrent neural network with behaviors governed by the density model. Auto numerical differentiation gives you the gradient for model optimization.
